### PR TITLE
Add and use BCP-47 time-zone identifiers in TimeZonesProvider

### PR DESCRIPTION
**Pull request by samchen61661**
Add and use BCP-47 time-zone identifiers in TimeZonesProvider (https://github.com/unicode-org/icu4x/issues/606)

### CODE REVIEW

File: provider/testdata/data/cldr/cldr-bcp47/bcp47/timezone.json
Comment (sffc) [inline provider/testdata/data/cldr/cldr-bcp47/bcp47/timezone.json:1]: New JSON files should be added to testdata/Cargo.toml so that they are automatically updated

**Review by sffc — COMMENTED** [review]
Looks correct so far!

**Review by nordzilla — COMMENTED** [review]
This is looking good. 

It looks like we're reading in the data, but not using it yet.

File: provider/cldr/src/transform/time_zones/convert.rs
Comment (sffc) [inline provider/cldr/src/transform/time_zones/convert.rs:109]: Suggestion, here and below: pull the key lookup out of this match and up after the `key.push_str(&inner_key);` on line 101

Comment (sffc) [inline provider/cldr/src/transform/time_zones/convert.rs:109]: Issue: If the mapping is not found, we should not fall back to `None`. I think we should probably return an error. For now you can panic.

Comment (samchen61661) [reply provider/cldr/src/transform/time_zones/convert.rs:109]: lines 117 & 118 also appends string in the `key`. If I pull the key lookup to line 101, the key lookup would return `None` for the branch `LocationOrSubRegion::SubRegion`.

**Review by sffc — COMMENTED** [review]
LGTM for what you have so far! I think you will need to update the testdata, `cargo make testdata`, and look at CI for other potential test failures.

File: provider/datagen/src/cldr/transform/time_zones/mod.rs
Comment (samchen61661) [inline provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: @sffc @Manishearth `time_zone_names` is a reference (not 'static or owned). Do you know if there is a better way to handle it instead of cloning? 

If I make struct `CldrTimeZonesData` take reference `time_zone_names`, it will raise error  '`self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement'.

Note that `cldr_time_zones_data` will be passed in `Yokeable::from` in line 103.

Comment (Manishearth) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: We don't care too much about clones in datagen, this is fine.

**Review by Manishearth — APPROVED** [review]
took a cursory look; this looks decent, but I'd wait for Shane/Erik's final approval

**Review by nordzilla — DISMISSED** [review]
This looks great to me, Sam.

Thanks so much!

File: provider/datagen/src/cldr/transform/time_zones/mod.rs
Comment (sffc) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: I suggest changing `CldrTimeZonesData` to be

```rust
pub struct CldrTimeZonesData<'a> {
    pub time_zone_names: &'a TimeZoneNames,
    pub bcp47_tzids: &'a LiteMap<String, String>,
}
```

and then you can save references.

> We don't care too much about clones in datagen, this is fine.

This is generally true, but this is effectively cloning a large object in a tight loop, which would be nice to avoid, since we have an easy way to avoid it.

**Review by sffc — CHANGES_REQUESTED** [review]
Looks great; just requesting the one change about lifetimes/cloning.

File: provider/datagen/src/cldr/transform/time_zones/mod.rs
Comment (sffc) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: Also, consider moving the struct to be inside the `transform` tree, rather than `cldr_serde`, since it isn't constructed or used anywhere else.

Comment (nordzilla) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: > I suggest changing `CldrTimeZonesData` to be
> 
> ```rust
> pub struct CldrTimeZonesData<'a> {
>     pub time_zone_names: &'a TimeZoneNames,
>     pub bcp47_tzids: &'a LiteMap<String, String>,
> }
> ```
> 
> and then you can save references.

Sam tried this and ran into issues that Yoke requires a static lifetime. 

The error that Sam mentioned above: 
```
`self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement.
```

Essentially, we can't guarantee that the lifetime of `Self`, to which the reference to `TimeZoneNames` belongs, would necessarily outlive the data payload that we create via:

```rust
Some(DataPayload::from_owned(<$marker as DataMarker>::Yokeable::from(cldr_time_zones_data))
```

I told Sam to not worry about the clones and that we could tackle this as a separate efficiency problem if it causes performance issues. 

The lifetime reasoning from the compiler seems sound to me. I think it is correct that `Self` would not necessarily outlive the data payload, but I'm less familiar with the internals of Yoke. 

@sffc if you have the solution for this, then we should apply it, but I didn't see one myself.

Comment (sffc) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: CldrTimeZonesData is never itself converted to a Yokeable. The From impl should clone the data, but only the small part of the data it needs.

Comment (nordzilla) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: It just seems to be that the data contained within `CldrTimeZonesData` needs to be owned, i.e. copied before the `From` impl is even invoked, otherwise it doesn't meet Yoke's `'static` lifetime requirements. 

It would be great to be able to pass a reference into the `From` impl, but it doesn't seem to be compatible with Yoke.

Am I misunderstanding something?

Comment (sffc) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: There are multiple things going on here:

```
Some(
    DataPayload::from_owned(
        <$marker as DataMarker>::Yokeable::from(
            cldr_time_zones_data
)))
```

`<$marker as DataMarker>::Yokeable::from` expands to something like `ExemplarCitiesV1<'static>::from`, and it can invoke a `From` impl such as

```rust
impl From<&CldrTimeZonesData<'_>> for ExemplarCitiesV1<'static>
```

Then, you call `DataPayload::from_owned(...)` which is the thing that requires the `'static`, but that's OK, because we have it since it is the return value of the `impl From`.

Comment (nordzilla) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: So @samchen61661 based on what Shane just said it looks like you may need to make the `From` impls take a reference to `CldrTimeZoneNames` instead of by value.

That is,

```rust
impl From<&CldrTimeZonesData<'_>> for ExemplarCitiesV1<'static>
```

instead of what it is currently:

```rust
impl From<CldrTimeZonesData<'_>> for ExemplarCitiesV1<'_>
```

Then hopefully things will work with references.

Comment (sffc) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs:95]: What you currently have is

```rust
impl From<CldrTimeZonesData> for TimeZoneFormatsV1<'_> {
```

You need to make it

```rust
impl From<&CldrTimeZonesData<'_>> for TimeZoneFormatsV1<'static> {
```

It doesn't actually matter whether `CldrTimeZonesData<>` is by value or reference. What matters is that the lifetime parameter on `CldrTimeZonesData<>` is not constrained to be the same as the lifetime parameter on `TimeZoneFormatsV1<>`.

Comment (samchen61661) [reply provider/datagen/src/cldr/transform/time_zones/mod.rs]: great! It is working now!

**Review by sffc — APPROVED** [review]

**Review by nordzilla — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**
